hivemind/dht/routing.py

```diff
@@ -93,41 +108,45 @@ def get_nearest_neighbors(
         :param query_id: find neighbors of this node
         :param k: find this many neighbors. If there aren't enough nodes in the table, returns all nodes
         :param exclude: if True, results will not contain query_node_id even if it is in table
-        :returns: a list of tuples (node_id, endpoint) for up to k neighbors sorted from nearest to farthest
-
-        :note: this is a semi-exhaustive search of nodes that takes O(n * log k) time.
-            One can implement a more efficient knn search using a binary skip-tree in some
-            more elegant language such as c++ / cython / numba.
-            Here's a sketch
-
-            Preparation: construct a non-regular binary tree of depth (2 * DHTID.HASH_NBYTES)
-             Each leaf corresponds to a binary DHTID with '0' for every left turn and '1' for right turn
```

Comment:
note: instead of traversing a skip-tree of nodes, we traverse a regular tree of buckets and select best nodes in buckets using heap-sort. This is 90% as fast (speculation!) but much simpler in terms of code